File: implementation/savingsgoal/src/main/java/com/financetracker/savingsgoal/logic/model/PeriodicalSavingsGoal.java
package com.financetracker.savingsgoal.logic.model;

import jakarta.persistence.*;
import lombok.*;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Builder(builderMethodName = "with")
@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "periodical_savings_goal")
public class PeriodicalSavingsGoal {

    @Id
    private UUID id;

    @Column(name = "description")
    private String description;

    @Column(name = "name")
    private String name;

    @Column(name = "achievement_status")
    private com.financetracker.savingsgoal.logic.model.AchievementStatus achievementStatus;

    @Column(name = "source_bank_account_id")
    private UUID sourceBankAccountId;

    @Column(name = "target_bank_account_id")
    private UUID targetBankAccountId;

    @Embedded
    @AttributeOverride(name = "amount", column = @Column(name = "goal_amount"))
    private com.financetracker.savingsgoal.logic.model.MonetaryAmount goal;

    @Column(name = "recurring_rate")
    private Double recurringRate;

    @Embedded
    @AttributeOverride(name = "amount", column = @Column(name = "recurring_amount_amount"))
    private com.financetracker.savingsgoal.logic.model.MonetaryAmount recurringAmount;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "start", column = @Column(name = "start_date")),
            @AttributeOverride(name = "end", column = @Column(name = "end_date"))
    })
    private com.financetracker.savingsgoal.logic.model.Duration duration;

    @Column(name = "periodicity")
    private com.financetracker.savingsgoal.logic.model.Periodicity periodicity;

    @Setter
    @JoinColumn(name = "savings_goal_id", referencedColumnName = "id")
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private Set<com.financetracker.savingsgoal.logic.model.SavingsRecord> savingsRecords = new HashSet<>();

}

File: implementation/ui-financetracker/src/app/savings-goal-periodical/periodicalSavingsGoal.ts
import { MonetaryAmount } from "../product/product";
import { AchievementStatus } from "../savings-goal-rule-based/ruleBasedSavingsGoal";
import { Periodicity } from "../transaction/transaction";

export class PeriodicalSavingsGoal {
    id: string;
    name: string;
    description: string;
    achievementStatus: AchievementStatus;
    sourceBankAccountId: string;
    targetBankAccountId: string;
    goal: MonetaryAmount;
    recurringRate: number | null;
    recurringAmount: MonetaryAmount;
    duration: string;
    periodicity: Periodicity;
    savingsRecords: SavingsRecord[]


    constructor(id: string, name: string,
                description: string,
                achievementStatus: AchievementStatus,
                sourceBankAccountId: string,
                targetBankAccountId: string,
                goal: MonetaryAmount,
                recurringRate: number,
                recurringAmount: MonetaryAmount,
                duration: string,
                periodicity: Periodicity,
                savingsRecords: SavingsRecord[]) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.achievementStatus = achievementStatus;
        this.sourceBankAccountId = sourceBankAccountId;
        this.targetBankAccountId = targetBankAccountId;
        this.goal = goal;
        this.recurringRate = recurringRate;
        this.recurringAmount = recurringAmount;
        this.duration = duration;
        this.periodicity = periodicity;
        this.savingsRecords = savingsRecords;
    }
}

export class SavingsRecord {
    id: string;
    date: string;
    amount: MonetaryAmount
    achievementStatus: AchievementStatus;

    constructor(id: string, date: string, amount: MonetaryAmount, achievementStatus: AchievementStatus) {
        this.id = id;
        this.date = date;
        this.amount = amount;
        this.achievementStatus = achievementStatus;
    }
}
